### PR TITLE
Store and restore hash from a compiled module

### PR DESCRIPTION
**Pull request by maminrayej**
This PR stores the module hash in the `ModuleInfo` as part of the `Artifact`, so later after loading the compiled module from the disk, we would have access to the module hash.

### CODE REVIEW

**Review by syrusakbary — CHANGES_REQUESTED** [review]
Move `ModuleHash` into `wasmer-types` and reuse it here as well

File: lib/api/src/js/module.rs [hunk 1]
```diff
@@ -42,6 +42,7 @@ pub struct Module {
     type_hints: Option<ModuleTypeHints>,
     #[cfg(feature = "js-serializable-module")]
     raw_bytes: Option<Bytes>,
+    hash: [u8; 8],
```
Comment (syrusakbary) [inline lib/api/src/js/module.rs:45]: In JS we should not need the hash, compiledModule is not usable in JS

File: lib/types/src/module.rs [hunk 2]
```diff
@@ -111,6 +111,10 @@ pub struct ModuleInfo {
     #[cfg_attr(feature = "enable-serde", serde(skip_serializing, skip_deserializing))]
     pub id: ModuleId,
 
+    // FIXME: We need ModuleHash here
```
Comment (syrusakbary) [inline lib/types/src/module.rs:114]: Please add a Rust comment on `ModuleInfo` (`///`).

Stating something like:

```
/// IMPORTANT: if you change this class, don't forget to update `MetadataHeader::version` to make sure
/// we don't break compatibility between versions
```

File: lib/types/src/module.rs [hunk 3]
```diff
@@ -111,6 +111,10 @@ pub struct ModuleInfo {
...
+    /// hash of the module
+    pub hash: [u8; 8],
```
Comment (syrusakbary) [inline lib/types/src/module.rs:116]: This needs to be `ModuleHash`

File: lib/api/Cargo.toml [hunk 4]
```diff
@@ -35,6 +35,7 @@ tracing = { version = "0.1" }
 wat = { version = "=1.0.71", optional = true }
 rustc-demangle = "0.1"
 shared-buffer = { workspace = true }
+xxhash-rust = { version = "0.8.10", features = ["xxh64"] }
```
Comment (syrusakbary) [inline lib/api/Cargo.toml:38]: We don't need this here

File: lib/api/src/js/module.rs [hunk 5]
```diff
@@ -210,6 +211,10 @@ impl Module {
         self.name.as_ref().map(|s| s.as_ref())
     }
 
+    pub fn hash(&self) -> Option<ModuleHash> {
+        None
+    }
+
```
Comment (syrusakbary) [inline lib/api/src/js/module.rs:217]: We don't need this changes

File: lib/api/src/jsc/module.rs [hunk 6]
```diff
@@ -13,6 +13,7 @@ use bytes::Bytes;
 use rusty_jsc::{JSObject, JSString, JSValue};
 use std::path::Path;
 use tracing::{debug, warn};
+use wasmer_types::ModuleHash;
```
Comment (syrusakbary) [inline lib/api/src/jsc/module.rs:16]: We don't need this changes

File: lib/api/src/jsc/module.rs [hunk 7]
```diff
@@ -175,6 +177,10 @@ impl Module {
         self.name.as_ref().map(|s| s.as_ref())
     }
 
+    pub fn hash(&self) -> Option<ModuleHash> {
+        self.info.hash
+    }
+
```
Comment (syrusakbary) [inline lib/api/src/jsc/module.rs:183]: We don't need this changes

File: lib/api/src/module.rs [hunk 8]
```diff
@@ -3,6 +3,7 @@ use std::fmt;
 use std::fs;
 use std::io;
 use std::path::Path;
+use wasmer_types::ModuleHash;
```
Comment (syrusakbary) [inline lib/api/src/module.rs:6]: We don't need this

File: lib/api/src/module.rs [hunk 9]
```diff
@@ -366,6 +367,11 @@ impl Module {
         self.0.name()
     }
 
+    /// Returns the xxhash module hash
+    pub fn hash(&self) -> Option<ModuleHash> {
+        self.0.hash()
+    }
+
     /// Sets the name of the current module.
```
Comment (syrusakbary) [inline lib/api/src/module.rs:375]: We don't need this

File: lib/api/src/module_info_polyfill.rs [hunk 10]
```diff
@@ -253,6 +253,8 @@ fn transform_err(err: BinaryReaderError) -> String {
 pub fn translate_module<'data>(data: &'data [u8]) -> WasmResult<ModuleInfoPolyfill> {
     let mut module_info: ModuleInfoPolyfill = Default::default();
 
+    module_info.info.hash = Some(ModuleHash::xxhash(data));
+
```
Comment (syrusakbary) [inline lib/api/src/module_info_polyfill.rs:257]: We don;t' need this

File: lib/api/src/sys/module.rs [hunk 11]
```diff
@@ -174,6 +175,10 @@ impl Module {
         self.info().name.as_deref()
     }
 
+    pub(crate) fn hash(&self) -> Option<ModuleHash> {
+        self.info().hash
+    }
+
```
Comment (syrusakbary) [inline lib/api/src/sys/module.rs:181]: We dont need this here

File: lib/types/src/module.rs [hunk 12]
```diff
@@ -97,8 +101,131 @@ mod serde_imports {
     }
 }
 
+/// The hash of a WebAssembly module.
```
Comment (syrusakbary) [inline lib/types/src/module.rs:104]: Can you move this into it's own file (`module_hash.rs`) ?

File: lib/cli/src/commands/run/mod.rs [hunk 13]
```diff
@@ -712,10 +711,9 @@ impl ExecutableTarget {
                 let engine = runtime.engine();
                 pb.set_message("Deserializing pre-compiled WebAssembly module");
                 let module = unsafe { Module::deserialize_from_file(&engine, path)? };
-                let module_hash = {
-                    let wasm = std::fs::read(path)?;
-                    ModuleHash::xxhash(wasm)
-                };
+
+                // FIXME: what if the artifact does not have the hash
+                let module_hash = module.hash().unwrap_or(ModuleHash::XXHash([0u8; 8]));
```
Comment (syrusakbary) [inline lib/cli/src/commands/run/mod.rs:716]: You should just do `module.info().hash()`

File: lib/compiler/src/translator/module.rs [hunk 14]
```diff
@@ -22,6 +22,8 @@ pub fn translate_module<'data>(
 ) -> WasmResult<ModuleTranslationState> {
     let mut module_translation_state = ModuleTranslationState::new();
 
+    environ.module.hash = Some(ModuleHash::xxhash(data));
+
```
Comment (syrusakbary) [inline lib/compiler/src/translator/module.rs:26]: The hash should not be set here, removeme

**Review by syrusakbary — CHANGES_REQUESTED** [review]

File: lib/api/src/js/module.rs [hunk 15]
```diff
@@ -115,7 +115,7 @@ impl Module {
             type_hints,
             name,
             #[cfg(feature = "js-serializable-module")]
-            raw_bytes: Some(binary),
+            raw_bytes: Some(binary.clone()),
```
Comment (syrusakbary) [inline lib/api/src/js/module.rs:118]: Not needed

File: lib/compiler/src/translator/environ.rs [hunk 16]
```diff
@@ -89,6 +89,9 @@ impl<'data> ModuleEnvironment<'data> {
         assert!(self.module_translation_state.is_none());
         let module_translation_state = translate_module(data, &mut self)?;
         self.module_translation_state = Some(module_translation_state);
+
+        self.module.hash = Some(ModuleHash::xxhash(data));
```
Comment (syrusakbary) [inline lib/compiler/src/translator/environ.rs:93]: We chatted about this, how are we solving the issue for not computing the hash always? This is not addressed

File: lib/api/src/engine.rs [hunk 17]
```diff
@@ -38,6 +38,12 @@ impl Engine {
         self.0.deterministic_id()
     }
 
+    /// Sets the hash algorithm
```
Comment (syrusakbary) [inline lib/api/src/engine.rs:41]: This needs to live in `NativeEngineExt`, not here

Comment (syrusakbary) [reply lib/api/src/engine.rs:41]: (remove it from here)

Comment (maminrayej) [reply lib/api/src/engine.rs:41]: You're right!

File: lib/api/src/sys/engine.rs [hunk 18]
```diff
@@ -55,7 +55,12 @@ pub(crate) fn default_engine() -> Engine {
 pub trait NativeEngineExt {
     /// Create a new `Engine` with the given config
     #[cfg(feature = "compiler")]
-    fn new(compiler_config: Box<dyn CompilerConfig>, target: Target, features: Features) -> Self;
+    fn new(
```
Comment (syrusakbary) [inline lib/api/src/sys/engine.rs:58]: Undo

File: lib/api/src/sys/engine.rs [hunk 19]
```diff
@@ -104,8 +109,18 @@ pub trait NativeEngineExt {
 
 impl NativeEngineExt for crate::engine::Engine {
     #[cfg(feature = "compiler")]
-    fn new(compiler_config: Box<dyn CompilerConfig>, target: Target, features: Features) -> Self {
-        Self(Engine::new(compiler_config, target, features))
+    fn new(
```
Comment (syrusakbary) [inline lib/api/src/sys/engine.rs:115]: Undo, breaking change

File: lib/api/src/sys/tunables.rs [hunk 20]
```diff
@@ -272,7 +272,12 @@ mod tests {
 
         let tunables = TinyTunables {};
         #[allow(deprecated)]
-        let mut engine = Engine::new(compiler.into(), Default::default(), Default::default());
+        let mut engine = Engine::new(
+            compiler.into(),
+            Default::default(),
+            Default::default(),
+            None,
```
Comment (syrusakbary) [inline lib/api/src/sys/tunables.rs:279]: Undo

**Review by syrusakbary — CHANGES_REQUESTED** [review]

File: lib/types/src/module_hash.rs [hunk 21]
```diff
@@ -0,0 +1,135 @@
...
+        Self::xxhash_from_bytes(key)
+    }
+
+    /// Creates a random sha256 hash for the module
+    pub fn sha256_random() -> Self {
+        let mut rand = rand::thread_rng();
+        let mut key = [0u8; 32];
+        rand.fill_bytes(&mut key);
+        Self::sha256_from_bytes(key)
+    }
```
Comment (syrusakbary) [inline lib/types/src/module_hash.rs:67]: Can we remove this methods? They are not used anywhere, and it will help us to remove the dependency on rand

File: lib/types/Cargo.toml [hunk 22]
```diff
@@ -23,6 +23,17 @@ enum-iterator = "0.7.0"
 target-lexicon = { version = "0.12.2", default-features = false }
 enumset.workspace = true
 bytecheck = "0.6.8"
+rand = "0.8"
```
Comment (syrusakbary) [inline lib/types/Cargo.toml:26]: We can probably remove this, once we remove the methods

File: lib/types/src/module_hash.rs (hunk 21 above)
Comment (syrusakbary) [reply lib/types/src/module_hash.rs:67]: Correction: `ModuleHash::xxhash_random` is used only in wasix.
Lets move it just to wasix.

It will help slim down dependencies for wasmer-types and wasmer-js

File: lib/api/src/jsc/module.rs [hunk 23]
```diff
@@ -67,8 +67,9 @@ impl Module {
     ///
     pub(crate) unsafe fn from_js_module(module: JSObject, binary: impl IntoBytes) -> Self {
         let binary = binary.into_bytes();
+
         // The module is now validated, so we can safely parse it's types
-        let info = crate::module_info_polyfill::translate_module(&binary[..])
+        let mut info = crate::module_info_polyfill::translate_module(&binary[..])
```
Comment (syrusakbary) [inline lib/api/src/jsc/module.rs:72]: Undo this changes? (sure will make linting fail)

**Review by syrusakbary — APPROVED** [review]
LGTM

File: lib/cli/src/commands/inspect.rs [hunk 24]
```diff
@@ -27,6 +31,14 @@ impl Inspect {
 
     fn inner_execute(&self) -> Result<()> {
         let (store, _compiler_type) = self.store.get_store()?;
+
+        #[cfg(feature = "sys")]
+        {
+            let mut engine = store.engine().clone();
+            let hash_algorithm = self.hash_algorithm.unwrap_or_default().into();
+            engine.set_hash_algorithm(Some(hash_algorithm));
+        }
```
Comment (syrusakbary) [inline lib/cli/src/commands/inspect.rs:40]: We can skip this

File: lib/cli/src/commands/validate.rs [hunk 25]
```diff
@@ -25,6 +29,14 @@ impl Validate {
     }
     fn inner_execute(&self) -> Result<()> {
         let (store, _compiler_type) = self.store.get_store()?;
+
+        #[cfg(feature = "sys")]
+        {
+            let mut engine = store.engine().clone();
+            let hash_algorithm = self.hash_algorithm.unwrap_or_default().into();
+            engine.set_hash_algorithm(Some(hash_algorithm));
+        }
```
Comment (syrusakbary) [inline lib/cli/src/commands/validate.rs:38]: We can remove this
